File: Numeros Pseudo-Aleatorios/Multiply With Carry/MultiplyWithCarry.pas
{ 
 PRNG .......: Multiply With Carry (MWC)
 Ano ........: Desconhecido
 Autor ......: George Marsaglia
}

program CMWC_RND;

uses
  SysUtils, DateUtils;
  
type
  CmwcRnd = record
    randomValues: array[0..4095] of LongWord;
    carry: LongWord;
    currentIndex: Cardinal;
  end;

function log2ibase(n, acc: Integer): Integer;
begin
  if n = 1 then
    log2ibase := acc
  else
    log2ibase := log2ibase(n shr 1, acc + 1);
end;

function log2i(n: Integer): Integer;
begin
  log2i := log2ibase(n, 0);
end;

function rand32: LongWord;
var
  seed: LongWord = 0;
  result_: LongWord = 0;
  bits, remainingBits: Integer;
begin
  bits := log2i(MAXINT) + 1;
  remainingBits := 32 - bits;
  repeat
    result_ := (result_ shl bits) or ((seed * 1103515245 + 12345) shr (32 - bits));
    remainingBits := remainingBits - bits;
  until remainingBits <= 0;
  rand32 := result_;
end;

procedure initCMWC(var state: CmwcRnd; seed: Cardinal);
var
  i: Integer;
begin
  for i := 0 to 4095 do
  begin
    seed := (seed * 1103515245 + 12345) and $FFFFFFFF;
    state.randomValues[i] := seed;
  end;
  repeat
  begin
    seed := (seed * 1103515245 + 12345) and $FFFFFFFF;
    state.carry := seed;
  end;
  until state.carry < 809430660;
  state.currentIndex := 4095;
end;

function randCMWC(var state: CmwcRnd): LongWord;
var
  t: UInt64;
  x: LongWord;
begin
  state.currentIndex := (state.currentIndex + 1) and 4095;
  t := 18782 * UInt64(state.randomValues[state.currentIndex]) + state.carry;
  state.carry := t shr 32;
  x := t + state.carry;
  if x < state.carry then
  begin
    x := x + 1;
    state.carry := state.carry + 1;
  end;
  state.randomValues[state.currentIndex] := $FFFFFFFE - x;
  randCMWC := state.randomValues[state.currentIndex];
end;

var
  cmwc1, cmwc2: CmwcRnd;
  Xn1, Xn2: Cardinal;
  i: Integer;

begin

  Xn1 := 1234;
  Xn2 := LongWord(DateTimeToUnix(Now));  
  
  initCMWC(cmwc1, Xn1);
  initCMWC(cmwc2, Xn2);

  writeln('GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)');
  writeln('PSEUDORANDOM NUMBER GENERATOR (PRNG)');
  writeln('PADRAO: Multiply With Carry (MWC)');
  writeln('==========================================');
  writeln;
  writeln('    Semente: 1234        Semente: TIME');
  writeln('    -------------        -------------');

  for i := 1 to 10 do
    writeln(randCMWC(cmwc1):17, '        ', randCMWC(cmwc2):13);
    
end.

File: Numeros Pseudo-Aleatorios/Multiply With Carry/MultiplyWithCarry.c
/*
 PRNG .......: Multiply With Carry (MWC)
 Ano ........: Desconhecido
 Autor ......: George Marsaglia
*/

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int log2ibase(int n, int acc) {
  if (n == 1)
    return acc;
  else
    return log2ibase(n >> 1, acc + 1);
}

int log2i(int n) {
  return log2ibase(n, 0);
}

struct CmwcRnd {
  uint32_t randomValues[4096];
  uint32_t carry;
  unsigned currentIndex;
};

uint32_t rand32(void) {
  static uint32_t seed = 0;
  uint32_t result = 0;
  int bits = log2i(RAND_MAX) + 1;
  int remaining_bits = 32 - bits;
  do {
    result = result << bits | ((seed = (seed * 1103515245 + 12345) & 0xFFFFFFFF) >> (32 - bits));
    remaining_bits -= bits;
  } while (! (remaining_bits <= 0));
  return result;
}

void initCMWC(struct CmwcRnd *state, unsigned int seed) {
  int i;
  for (i = 0; i <= 4095; i++) {
    seed = (seed * 1103515245 + 12345) & 0xFFFFFFFF;
    state->randomValues[i] = seed;
  }
  do {
    seed = (seed * 1103515245 + 12345) & 0xFFFFFFFF;
    state->carry = seed;
  } while (! (state->carry < 809430660));
  state->currentIndex = 4095;
}

uint32_t randCMWC(struct CmwcRnd *state) {
  uint64_t t;
  uint32_t x;
  state->currentIndex = (state->currentIndex + 1) & 4095;
  t = 18782ULL * state->randomValues[state->currentIndex] + state->carry;
  state->carry = t >> 32;
  x = t + state->carry;
  if (x < state->carry) {
    x++;
    state->carry++;
  }
  return state->randomValues[state->currentIndex] = 0xfffffffe - x;
}

struct CmwcRnd cmwc1, cmwc2;
unsigned int Xn1, Xn2;
int i;
  
int main(void) {

  Xn1 = 1234;
  Xn2 = time(NULL);
    
  initCMWC(&cmwc1, Xn1);
  initCMWC(&cmwc2, Xn2);

  printf("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)\n");
  printf("PSEUDORANDOM NUMBER GENERATOR (PRNG)\n");
  printf("PADRAO: Multiply With Carry (MWC)\n");
  printf("==========================================\n\n");
  printf("    Semente: 1234        Semente: TIME\n");
  printf("    -------------        -------------\n");

  for (i = 1; i <= 10; i++)
    printf("%17u        %13u\n", randCMWC(&cmwc1), randCMWC(&cmwc2));
 
  return 0;
}
